RFC/src/RFC-0121_ConsensusEncoding.md

```diff
@@ -0,0 +1,347 @@
+3. `prev_hash` - fixed 32-bytes 
+4. `timestamp` - UNIX timestamp converted to a fixed 8-byte array (little endian) 
+5. `input_mr` - fixed 32-bytes
+6. `output_mmr_size` - `u64` converted to a fixed 8-byte array (little endian)
+7. `witness_mr` - fixed 32-bytes
+8. `kernel_mr` - fixed 32-bytes
+9. `kernel_mmr_size` - `u64` converted to a fixed 8-byte array (little endian)
+10. `total_kernel_offset` - 32-byte Scalar, see [RistrettoPrivateKey]
+11. `total_script_offset` - 32-byte Scalar, see [RistrettoPrivateKey]
+
+This pre-image is hashed and block hash is constructed, in order, as follows:
+
+1. `merge_mining_hash` - As above
+2. `pow_algo` - enumeration of types of PoW as a single unsigned byte, where `Monero = 0x00` and `Sha3 = 0x01`
+3. `pow_data` - raw variable bytes (no length varint)
+4. `nonce` - the PoW nonce, `u64` converted to a fixed 8-byte array (little endian)
+
+#### Output Features 
+[Output Features]: #output-metadata-and-features "Output Features"
+
+```rust,ignore
+pub struct OutputFeatures {
+  pub version: OutputFeaturesVersion,
+  pub maturity: u64,
+  pub flags: OutputFlags,
+  pub metadata: Vec<u8>,
+  pub unique_id: Option<Vec<u8>>,
+  pub parent_public_key: Option<PublicKey>,
+  pub asset: Option<AssetOutputFeatures>,
+  pub mint_non_fungible: Option<MintNonFungibleFeatures>,
+  pub sidechain_checkpoint: Option<SideChainCheckpointFeatures>,
+}
+```
+
+Output features consensus encoding is defined as follows (in order):
+
+1. `version` - 1 unsigned byte. This should always be `0x00` but is reserved for future proofing.
+2. `maturity` - [varint]
+3. `flags` - 1 unsigned byte
+4. `metadata` - [dynamic vector]
+5. `unique_id` - [nullable] + [dynamic vector]
+6. `parent_public_key` - [nullable] + 32-byte compressed public key 
+7. `asset` - [nullable] + [AssetOutputFeatures](#assetoutputfeatures)
+8. `mint_non_fungible` - [nullable] + [MintNonFungibleFeatures](#mintnonfungiblefeatures)
+9. `sidechain_checkpoint` - [nullable] + [SideChainCheckpointFeatures](#sidechaincheckpointfeatures)
+
+##### AssetOutputFeatures
+
+- `public_key` - [RistrettoPublicKey] 
+- `template_ids` - [dynamic vector] + [varint]
+- `template_parameters` - [dynamic vector]
+
+##### MintNonFungibleFeatures
+
+- `asset_public_key` - [RistrettoPublicKey] 
+- `asset_owner_commitment` - [RistrettoPublicKey] 
+
+##### SideChainCheckpointFeatures
+
+- `merkle_root` - [fixed sized array]
+- `committee` - [dynamic vector] + [RistrettoPublicKey]
+
+#### Transaction Output
+[Transaction Output]: #transaction-output "Transaction Output"
+
+```rust,ignore
+pub struct TransactionOutput {
+    pub version: u8,
+    pub features: OutputFeatures,
+    pub commitment: Commitment,
+    pub proof: RangeProof,
+    pub script: TariScript,
+    pub sender_offset_public_key: PublicKey,
+    pub metadata_signature: ComSignature,
+    pub covenant: Covenant,
+}
+```
+
+The canonical hash commits to the common data between an output and the input spending that output
```

Comment:
Why not use the same hash as the one in the output merkle tree?

Reply:
It is the same hash. Maybe the wording can be improved?